web/packages/teleterm/src/ui/DocumentCluster/resourcesContext.tsx

```diff
@@ -87,5 +93,24 @@ export const useResourcesContext = () => {
     );
   }
 
-  return context;
+  const {
+    requestResourcesRefresh: requestResourcesRefreshContext,
+    onResourcesRefreshRequest: onResourcesRefreshRequestContext,
+  } = context;
+
+  return {
+    requestResourcesRefresh: useCallback(
+      () => requestResourcesRefreshContext(rootClusterUri),
+      [requestResourcesRefreshContext, rootClusterUri]
+    ),
+    onResourcesRefreshRequest: useCallback(
+      (listener: () => void) =>
+        onResourcesRefreshRequestContext(requestedRootClusterUri => {
+          if (requestedRootClusterUri === rootClusterUri) {
+            listener();
+          }
+        }),
+      [onResourcesRefreshRequestContext, rootClusterUri]
+    ),
```

Comment:
What do you think about this implementation?

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleterm/src/ui/DocumentCluster/resourcesContext.tsx b/web/packages/teleterm/src/ui/DocumentCluster/resourcesContext.tsx
index 8cf37d4a82..a8794ba653 100644
--- a/web/packages/teleterm/src/ui/DocumentCluster/resourcesContext.tsx
+++ b/web/packages/teleterm/src/ui/DocumentCluster/resourcesContext.tsx
@@ -40,10 +40,12 @@ export interface ResourcesContext {
   requestResourcesRefresh: (rootClusterUri: RootClusterUri) => void;
   /**
    * onResourcesRefreshRequest registers a listener that will be called any time a refresh is
-   * requested. Typically called from useEffect, for this purpose it returns a cleanup function.
+   * requested for a particular rootClusterUri. Typically called from useEffect, for this purpose it
+   * returns a cleanup function.
    */
   onResourcesRefreshRequest: (
-    listener: (rootClusterUri: RootClusterUri) => void
+    rootClusterUri: RootClusterUri,
+    listener: () => void
   ) => {
     cleanup: () => void;
   };
@@ -64,7 +66,15 @@ export const ResourcesContextProvider: FC<PropsWithChildren> = props => {
   );
 
   const onResourcesRefreshRequest = useCallback(
-    (listener: (rootClusterUri: RootClusterUri) => void) => {
+    (
+      targetRootClusterUri: RootClusterUri,
+      listenerWithoutRootClusterUri: () => void
+    ) => {
+      const listener = (rootClusterUri: RootClusterUri) => {
+        if (rootClusterUri === targetRootClusterUri) {
+          listenerWithoutRootClusterUri();
+        }
+      };
       emitterRef.current.addListener('refresh', listener);
 
       return {
@@ -105,11 +115,7 @@ export const useResourcesContext = (rootClusterUri: RootClusterUri) => {
     ),
     onResourcesRefreshRequest: useCallback(
       (listener: () => void) =>
-        onResourcesRefreshRequestContext(requestedRootClusterUri => {
-          if (requestedRootClusterUri === rootClusterUri) {
-            listener();
-          }
-        }),
+        onResourcesRefreshRequestContext(rootClusterUri, listener),
       [onResourcesRefreshRequestContext, rootClusterUri]
     ),
   };

```
</details>

What we're doing with `useResourcesContext` is a rather new approach to how we structure our hooks, so I've been toying a bit with different ways to accomplish this.

The implementation from my patch makes it so that `useResourcesContext` is kept simple – all it does is remove the need to manually pass `rootClusterUri` by the caller. All the details of the implementation are then moved to `ResourcesContextProvider` which does stuff like inspecting `rootClusterUri`.